File: ios/Sources/AmthalPortal/BridgeController.swift
import Foundation
import WebKit

/// Internal event surface the bridge raises toward the host view controller.
@MainActor
protocol BridgeControllerDelegate: AnyObject {
    func bridgeDidReceiveHello(portalVersion: String?)
    func bridgeDidBecomeReady(portalVersion: String?)
    func bridgeAuthExpired(reason: String?)
    func bridgeFormSubmitted(_ result: FormSubmitResult)
    func bridgeFormCancelled()
    func bridgeNavigated(path: String, title: String?)
    func bridgeOpenExternal(urlString: String)
    func bridgeDownloadRequested(_ payload: DownloadRequestPayload)
    func bridgeHaptic(style: String)
    func bridgeHelloTimedOut()
}

/// Owns the message-level protocol: envelope parsing/serialization, the
/// hello→init handshake, request/ack correlation with timeouts, and the
/// origin allowlist checks required before any inbound dispatch or outbound
/// evaluation.
@MainActor
final class BridgeController: NSObject {
    /// WKScriptMessageHandler name — the web shim probes
    /// `window.webkit.messageHandlers.amthal`.
    static let handlerName = "amthal"

    weak var webView: WKWebView?
    weak var delegate: BridgeControllerDelegate?

    /// Latest credentials; re-sent in every `init` reply to `hello`.
    var currentAuth: PortalAuth

    /// Latest appearance configuration; re-sent in every `init`.
    var currentConfigure: ConfigurePayload

    private(set) var helloReceived = false

    private let config: PortalConfig
    private var helloTimeoutTask: Task<Void, Never>?
    private var pending: [String: PendingRequest] = [:]

    private struct PendingRequest {
        let completion: (Data?) -> Void
        let timeoutTask: Task<Void, Never>
    }

    init(config: PortalConfig, auth: PortalAuth) {
        self.config = config
        self.currentAuth = auth
        self.currentConfigure = ConfigurePayload(
            locale: config.locale.rawValue,
            theme: config.theme.rawValue,
            fontScale: config.fontScale
        )
        super.init()
    }

    // MARK: - Lifecycle

    /// Call when a main-frame load finishes: arms the 15 s `hello` watchdog.
    func pageDidFinishLoad() {
        guard !helloReceived else { return }
        helloTimeoutTask?.cancel()
        helloTimeoutTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(BridgeProtocol.helloTimeout * 1_000_000_000))
            guard !Task.isCancelled, let self, !self.helloReceived else { return }
            self.delegate?.bridgeHelloTimedOut()
        }
    }

    /// Call before every fresh page load and on teardown: cancels the
    /// watchdog and resolves all pending round-trips with their safe defaults.
    func resetForNewLoad() {
        helloReceived = false
        helloTimeoutTask?.cancel()
        helloTimeoutTask = nil
        let inFlight = pending
        pending.removeAll()
        for (_, request) in inFlight {
            request.timeoutTask.cancel()
            request.completion(nil)
        }
    }

    // MARK: - Outbound

    /// `backPressed` round-trip. Timeout / missing answer ⇒ `handled:false`.
    func sendBackPressed(completion: @escaping (_ handled: Bool) -> Void) {
        sendRequest(type: MessageType.backPressed) { data in
            let payload = data.flatMap { try? JSONDecoder().decode(BackHandledPayload.self, from: $0) }
            completion(payload?.handled ?? false)
        }
    }

    /// `dismissRequested` round-trip. Timeout / missing answer ⇒ `dirty:false`.
    func sendDismissRequested(completion: @escaping (_ dirty: Bool) -> Void) {
        sendRequest(type: MessageType.dismissRequested) { data in
            let payload = data.flatMap { try? JSONDecoder().decode(DismissStatePayload.self, from: $0) }
            completion(payload?.dirty ?? false)
        }
    }

    /// (Re)issues credentials. Web MUST ack; `completion` receives the ack
    /// payload, or nil on timeout.
    func sendAuth(completion: ((AckPayload?) -> Void)? = nil) {
        let payload = makeAuthPayload()
        if let completion {
            sendRequest(type: MessageType.auth, payload: payload) { data in
                completion(data.flatMap { try? JSONDecoder().decode(AckPayload.self, from: $0) })
            }
        } else {
            send(type: MessageType.auth, payload: payload)
        }
    }

    /// Re-applies locale/theme/fontScale at runtime.
    func sendConfigure() {
        send(type: MessageType.configure, payload: currentConfigure)
    }

    /// Teardown notification; the web clears its persisted auth on receipt.
    func sendDestroy() {
        send(type: MessageType.destroy, payload: Optional<EmptyPayload>.none)
    }

    // MARK: - Outbound plumbing

    private func send<P: Encodable>(
        type: String,
        payload: P?,
        id: String = UUID().uuidString,
        replyTo: String? = nil
    ) {
        let envelope = OutboundEnvelope(id: id, replyTo: replyTo, type: type, payload: payload)
        guard let data = try? JSONEncoder().encode(envelope) else {
            log("failed to encode envelope of type \(type)")
            return
        }
        evaluate(envelopeData: data, type: type)
    }

    private func sendRequest<P: Encodable>(
        type: String,
        payload: P? = Optional<EmptyPayload>.none,
        timeout: TimeInterval = BridgeProtocol.requestTimeout,
        completion: @escaping (Data?) -> Void
    ) {
        let id = UUID().uuidString
        let timeoutTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
            guard !Task.isCancelled, let self,
                  let request = self.pending.removeValue(forKey: id)
            else { return }
            self.log("request \(type) (\(id)) timed out after \(timeout)s")
            request.completion(nil)
        }
        pending[id] = PendingRequest(completion: completion, timeoutTask: timeoutTask)
        send(type: type, payload: payload, id: id)
    }

    /// Serializes the envelope to a JSON string `S`, then evaluates
    /// `window.__amthalReceive(<JSON-encode of S>)` — the string is itself
    /// JSON-quoted, guaranteeing correct escaping. Verifies the WebView's
    /// current origin against the allowlist before every evaluation.
    private func evaluate(envelopeData: Data, type: String) {
        guard let webView else { return }
        guard config.isOriginAllowed(webView.url) else {
            log("dropped outbound \(type): WebView origin not allowlisted")
            return
        }
        guard let json = String(data: envelopeData, encoding: .utf8),
              let quotedData = try? JSONSerialization.data(
                withJSONObject: json,
                options: [.fragmentsAllowed]
              ),
              let quoted = String(data: quotedData, encoding: .utf8)
        else {
            log("failed to quote envelope of type \(type)")
            return
        }
        webView.evaluateJavaScript("window.__amthalReceive(\(quoted));") { [weak self] _, error in
            if error != nil {
                self?.log("evaluateJavaScript failed for \(type)")
            }
        }
    }

    private func sendInit(replyTo: String) {
        let payload = InitPayload(
            sdkVersion: AmthalPortalSDK.version,
            platform: "ios",
            protocolVersion: BridgeProtocol.version,
            configure: currentConfigure,
            auth: makeAuthPayload()
        )
        send(type: MessageType.bridgeInit, payload: payload, replyTo: replyTo)
    }

    private func makeAuthPayload() -> AuthPayload {
        var user: JSONValue?
        if let userJSON = currentAuth.userJSON,
           let data = userJSON.data(using: .utf8) {
            user = try? JSONDecoder().decode(JSONValue.self, from: data)
            if user == nil {
                log("auth.userJSON is not valid JSON — omitting `user` from payload")
            }
        }
        return AuthPayload(token: currentAuth.token, branchId: currentAuth.branchId, user: user)
    }

    // MARK: - Inbound

    private func handle(_ message: InboundMessage) {
        // Correlate replies to pending requests first. A message with
        // `replyTo` must never be re-replied to.
        if let replyTo = message.replyTo,
           let request = pending.removeValue(forKey: replyTo) {
            request.timeoutTask.cancel()
            request.completion(message.payloadData)
            return
        }

        switch message.type {
        case MessageType.hello:
            helloReceived = true
            helloTimeoutTask?.cancel()
            helloTimeoutTask = nil
            // Native MUST answer every hello with init (idempotent).
            sendInit(replyTo: message.id)
            let hello = message.decodePayload(HelloPayload.self)
            delegate?.bridgeDidReceiveHello(portalVersion: hello?.portalVersion)

        case MessageType.ready:
            let ready = message.decodePayload(ReadyPayload.self)
            delegate?.bridgeDidBecomeReady(portalVersion: ready?.portalVersion)

        case MessageType.authExpired:
            let payload = message.decodePayload(AuthExpiredPayload.self)
            delegate?.bridgeAuthExpired(reason: payload?.reason)

        case MessageType.formSubmitted:
            if let result = message.decodePayload(FormSubmitResult.self) {
                delegate?.bridgeFormSubmitted(result)
            } else {
                log("formSubmitted payload could not be decoded — ignored")
            }

        case MessageType.formCancelled:
            delegate?.bridgeFormCancelled()

        case MessageType.navigate:
            if let payload = message.decodePayload(NavigatePayload.self) {
                delegate?.bridgeNavigated(path: payload.path, title: payload.title)
            }

        case MessageType.openExternal:
            if let payload = message.decodePayload(OpenExternalPayload.self) {
                delegate?.bridgeOpenExternal(urlString: payload.url)
            }

        case MessageType.downloadRequest:
            if let payload = message.decodePayload(DownloadRequestPayload.self) {
                delegate?.bridgeDownloadRequested(payload)
            }

        case MessageType.haptic:
            if let payload = message.decodePayload(HapticPayload.self) {
                delegate?.bridgeHaptic(style: payload.style)
            }

        case MessageType.log:
            // Diagnostics channel — MUST be dropped in release builds.
            #if DEBUG
            if config.enableLogging, let payload = message.decodePayload(LogPayload.self) {
                print("[AmthalPortal][web][\(payload.level ?? "log")] \(payload.message ?? "")")
            }
            #endif

        case MessageType.ack:
            // Stray/late ack with no pending request — ignore.
            break

        default:
            // Unknown types MUST be ignored (forward compatibility).
            break
        }
    }

    // MARK: - Logging

    private func log(_ message: String) {
        #if DEBUG
        if config.enableLogging {
            print("[AmthalPortal] \(message)")
        }
        #endif
    }
}

// MARK: - WKScriptMessageHandler

extension BridgeController: WKScriptMessageHandler {
    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard message.name == Self.handlerName else { return }

        // Origin gate 1 (normative): the WebView's current URL origin must be
        // allowlisted before any inbound message is dispatched.
        guard config.isOriginAllowed(webView?.url) else {
            log("dropped inbound message: WebView origin not allowlisted")
            return
        }

        // Origin gate 2 (defense in depth): the posting frame itself must
        // belong to an allowlisted origin.
        let frameHost = message.frameInfo.securityOrigin.host
        guard config.isHostAllowed(frameHost) else {
            log("dropped inbound message: frame origin not allowlisted")
            return
        }

        guard let inbound = InboundMessage(body: message.body) else {
            // Wrong `v`, malformed envelope, etc. — silently ignored per spec.
            return
        }
        handle(inbound)
    }
}

File: android/portal-sdk/src/main/java/com/amthalgroup/portal/internal/BridgeController.kt
package com.amthalgroup.portal.internal

import android.os.Handler
import android.os.Looper
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import com.amthalgroup.portal.PortalConfig
import org.json.JSONException
import org.json.JSONObject

/**
 * Bridge Protocol v1 endpoint on the native side.
 *
 * - Inbound: a single `@JavascriptInterface` object named `__amthalAndroid`
 *   (installed by [com.amthalgroup.portal.AmthalPortalView] BEFORE `loadUrl`).
 * - Outbound: `window.__amthalReceive(<JSON-quoted envelope string>)` via
 *   `evaluateJavascript`.
 *
 * Every inbound dispatch and every outbound evaluation first verifies that
 * the WebView's current URL origin is allowed (SPEC.md §2.3).
 * All WebView interaction happens on the main looper.
 */
internal class BridgeController(
    private val webView: WebView,
    private val config: PortalConfig,
    private val host: Host
) {

    internal interface Host {
        /** Current `configure` payload ({locale, theme, fontScale}). */
        fun buildConfigurePayload(): JSONObject

        /** Current `auth` payload ({token, branchId?, user?}). */
        fun buildAuthPayload(): JSONObject

        fun onHello(portalVersion: String)
        fun onReady(portalVersion: String)
        fun onAuthExpired(reason: String?)
        fun onFormSubmitted(payload: JSONObject?)
        fun onFormCancelled()
        fun onNavigate(path: String, title: String?)
        fun onOpenExternal(url: String)
        fun onDownloadRequest(payload: JSONObject)
        fun onBridgeTimeout()
    }

    private class Pending(val timeout: Runnable, val callback: (JSONObject?) -> Unit)

    private val mainHandler = Handler(Looper.getMainLooper())
    private val pending = HashMap<String, Pending>()
    private var helloReceivedThisPage = false
    private var readyEmittedThisPage = false
    private var helloTimeout: Runnable? = null
    private var shutDown = false

    /** The object registered as `window.__amthalAndroid`. */
    val jsInterface = JsBridge()

    internal inner class JsBridge {
        /** Called by the web shim on an arbitrary WebView-internal thread. */
        @JavascriptInterface
        fun postMessage(message: String?) {
            if (message == null) return
            mainHandler.post { handleInbound(message) }
        }
    }

    // ------------------------------------------------------------------ page lifecycle

    /** New main-frame page load started: reset per-page state, fail stale requests. */
    fun onPageStarted() {
        helloReceivedThisPage = false
        readyEmittedThisPage = false
        cancelHelloTimeout()
        failAllPending()
    }

    /** Page finished loading: expect `hello` within 15 s (unless it already arrived). */
    fun startHelloTimeout() {
        if (shutDown || helloReceivedThisPage) return
        cancelHelloTimeout()
        val r = Runnable {
            helloTimeout = null
            if (!helloReceivedThisPage && !shutDown) host.onBridgeTimeout()
        }
        helloTimeout = r
        mainHandler.postDelayed(r, BridgeProtocol.HELLO_TIMEOUT_MS)
    }

    // ------------------------------------------------------------------ outbound API

    /** (Re)issues credentials — `auth` message (web acks; ack is consumed silently). */
    fun sendAuth() {
        dispatch(BridgeProtocol.newEnvelope(BridgeProtocol.MessageType.AUTH, host.buildAuthPayload()))
    }

    /** Re-applies locale/theme/fontScale — `configure` message. */
    fun sendConfigure() {
        dispatch(BridgeProtocol.newEnvelope(BridgeProtocol.MessageType.CONFIGURE, host.buildConfigurePayload()))
    }

    /**
     * `backPressed` round-trip. [onResult] is invoked on the main thread with
     * the web's `{handled}` answer; timeout (10 s) counts as `handled = false`.
     */
    fun requestBackPressed(onResult: (handled: Boolean) -> Unit) {
        request(BridgeProtocol.MessageType.BACK_PRESSED, null, BridgeProtocol.REQUEST_TIMEOUT_MS) { reply ->
            onResult(reply?.optBoolean("handled", false) ?: false)
        }
    }

    /**
     * `dismissRequested` round-trip (dirty-state guard). Timeout counts as
     * `dirty = false`.
     */
    fun requestDismissState(onResult: (dirty: Boolean) -> Unit) {
        request(BridgeProtocol.MessageType.DISMISS_REQUESTED, null, BridgeProtocol.REQUEST_TIMEOUT_MS) { reply ->
            onResult(reply?.optBoolean("dirty", false) ?: false)
        }
    }

    /** Sends `destroy`; [onDone] fires on ack or after 1 s, whichever is first. */
    fun sendDestroy(onDone: () -> Unit) {
        request(BridgeProtocol.MessageType.DESTROY, null, BridgeProtocol.DESTROY_ACK_TIMEOUT_MS) { onDone() }
    }

    /** Stops timers and drops pending requests. No WebView access after this. */
    fun shutdown() {
        shutDown = true
        cancelHelloTimeout()
        for (p in pending.values) mainHandler.removeCallbacks(p.timeout)
        pending.clear()
    }

    // ------------------------------------------------------------------ internals

    /** Main thread only. */
    private fun handleInbound(message: String) {
        if (shutDown) return
        if (!isCurrentOriginAllowed()) {
            log("Dropped inbound bridge message from disallowed origin")
            return
        }
        val envelope = try {
            JSONObject(message)
        } catch (e: JSONException) {
            log("Dropped unparseable bridge message")
            return
        }

        // Wrong protocol version -> silently ignore (SPEC.md §1).
        if (envelope.optInt("v", -1) != BridgeProtocol.PROTOCOL_VERSION) return

        val type = envelope.optString("type", "")
        val id = envelope.optString("id", "")
        val replyTo = envelope.optStringOrNull("replyTo")
        val payload = envelope.optJSONObject("payload")

        if (replyTo != null) {
            val p = pending.remove(replyTo)
            if (p != null) {
                mainHandler.removeCallbacks(p.timeout)
                p.callback(payload)
                return
            }
            // A reply nobody is waiting for: never re-reply; acks end here.
            if (type == BridgeProtocol.MessageType.ACK) return
        }

        when (type) {
            BridgeProtocol.MessageType.HELLO -> {
                helloReceivedThisPage = true
                cancelHelloTimeout()
                // Answer EVERY hello with init (idempotent on the web side).
                sendInit(replyToHelloId = id)
                host.onHello(payload?.optString("portalVersion").orEmpty())
            }
            BridgeProtocol.MessageType.READY -> {
                // `ready` is emitted once per page load; guard against repeats.
                if (!readyEmittedThisPage) {
                    readyEmittedThisPage = true
                    host.onReady(payload?.optString("portalVersion").orEmpty())
                }
            }
            BridgeProtocol.MessageType.AUTH_EXPIRED ->
                host.onAuthExpired(payload?.optStringOrNull("reason"))
            BridgeProtocol.MessageType.FORM_SUBMITTED ->
                host.onFormSubmitted(payload)
            BridgeProtocol.MessageType.FORM_CANCELLED ->
                host.onFormCancelled()
            BridgeProtocol.MessageType.NAVIGATE ->
                host.onNavigate(
                    payload?.optString("path").orEmpty(),
                    payload?.optStringOrNull("title")
                )
            BridgeProtocol.MessageType.OPEN_EXTERNAL -> {
                val url = payload?.optString("url").orEmpty()
                if (url.isNotEmpty()) host.onOpenExternal(url)
            }
            BridgeProtocol.MessageType.DOWNLOAD_REQUEST -> {
                if (payload != null) host.onDownloadRequest(payload)
            }
            BridgeProtocol.MessageType.HAPTIC -> {
                // Optional nicety; intentionally ignored on Android.
            }
            BridgeProtocol.MessageType.LOG -> {
                // Diagnostics channel — dropped unless logging is explicitly enabled.
                if (config.enableLogging) {
                    Log.d(TAG, "[web:${payload?.optString("level")}] ${payload?.optString("message")}")
                }
            }
            else -> {
                // Unknown type -> ignore (forward compatibility, SPEC.md §1).
            }
        }
    }

    private fun sendInit(replyToHelloId: String) {
        val payload = JSONObject()
        payload.put("sdkVersion", BridgeProtocol.SDK_VERSION)
        payload.put("platform", BridgeProtocol.PLATFORM)
        payload.put("protocolVersion", BridgeProtocol.PROTOCOL_VERSION)
        payload.put("configure", host.buildConfigurePayload())
        payload.put("auth", host.buildAuthPayload())
        dispatch(
            BridgeProtocol.newEnvelope(
                BridgeProtocol.MessageType.INIT,
                payload,
                replyTo = if (replyToHelloId.isEmpty()) null else replyToHelloId
            )
        )
    }

    private fun request(
        type: String,
        payload: JSONObject?,
        timeoutMs: Long,
        onReply: (JSONObject?) -> Unit
    ) {
        if (shutDown) {
            onReply(null)
            return
        }
        val envelope = BridgeProtocol.newEnvelope(type, payload)
        val id = envelope.getString("id")
        val timeout = Runnable {
            if (pending.remove(id) != null) onReply(null)
        }
        pending[id] = Pending(timeout, onReply)
        mainHandler.postDelayed(timeout, timeoutMs)
        dispatch(envelope)
    }

    /** Serializes + evaluates on the main thread, after the origin check. */
    private fun dispatch(envelope: JSONObject) {
        val run = Runnable {
            if (shutDown && envelope.optString("type") != BridgeProtocol.MessageType.DESTROY) return@Runnable
            if (!isCurrentOriginAllowed()) {
                log("Suppressed outbound '${envelope.optString("type")}' — WebView origin not allowed")
                return@Runnable
            }
            val js = BridgeProtocol.RECEIVE_FUNCTION + "(" + JSONObject.quote(envelope.toString()) + ")"
            webView.evaluateJavascript(js, null)
        }
        if (Looper.myLooper() == Looper.getMainLooper()) run.run() else mainHandler.post(run)
    }

    /** Main thread only (reads WebView.getUrl()). */
    private fun isCurrentOriginAllowed(): Boolean = config.isUrlOriginAllowed(webView.url)

    private fun failAllPending() {
        if (pending.isEmpty()) return
        val stale = pending.values.toList()
        pending.clear()
        for (p in stale) {
            mainHandler.removeCallbacks(p.timeout)
            p.callback(null)
        }
    }

    private fun cancelHelloTimeout() {
        helloTimeout?.let { mainHandler.removeCallbacks(it) }
        helloTimeout = null
    }

    private fun log(message: String) {
        if (config.enableLogging) Log.d(TAG, message)
    }

    private companion object {
        private const val TAG = "AmthalPortal"
    }
}
